### PR TITLE
Improved optional perf and changed build scripts so that LiteNetLib has proper directives

### PR DESCRIPTION
**Pull request by Measurity**
- NitroxModel is now seperated from UnityEngine*.dll files by consistently using the custom data structures provided by NitroxModel.dll
 - Will be much easier to deploy Nitrox now and to make further changes to how Nitrox loads into Subnautica.

TODO:
 - Fix reflection caching for Unity object check
 - Fix 1 unit test that broke slightly
 - Resolve conflicts..

### CODE REVIEW

File: NitroxClient/GameLogic/Item.cs [hunk 1]
```diff
@@ -1,12 +1,12 @@
 ﻿using NitroxClient.Communication.Abstract;
 using NitroxClient.GameLogic.Helper;
 using NitroxClient.MonoBehaviours;
-using NitroxModel.DataStructures;
 using NitroxModel.DataStructures.Util;
 using NitroxModel.Logger;
 using NitroxModel.Packets;
-using NitroxModel_Subnautica.Helper;
+using NitroxModel_Subnautica.DataStructures;
```
Comment (Sunrunner37) [inline NitroxClient/GameLogic/Item.cs:7]: Instead, can we re-namespace unity/game surrogates?  This makes it harder to work with common structures such as NitroxId below.

File: NitroxClient/GameLogic/InitialSync/VehicleInitialSyncProcessor.cs [hunk 2]
```diff
@@ -23,11 +24,11 @@ public VehicleInitialSyncProcessor(Vehicles vehicles)
         public override IEnumerator Process(InitialPlayerSync packet, WaitScreen.ManualWaitItem waitScreenItem)
         {
             int totalSyncedVehicles = 0;
-            int nonCyclopsVehicleCount = packet.Vehicles.Where(v => v.TechType.Enum() != TechType.Cyclops).Count();
+            int nonCyclopsVehicleCount = packet.Vehicles.Count(v => v.TechType.ToUnity() != TechType.Cyclops);
 
             foreach (VehicleModel vehicle in packet.Vehicles)
             {
-                if (vehicle.TechType.Enum() != TechType.Cyclops)
+                if (vehicle.TechType.ToUnity() != TechType.Cyclops)
```
Comment (Sunrunner37) [inline NitroxClient/GameLogic/InitialSync/VehicleInitialSyncProcessor.cs:31]: Would it be easier to define == and != for our version of tech type?  Could we also define any implicit conversion operators for these items to make it less painful?

File: NitroxModel-Subnautica/DataStructures/DataExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,80 @@
+﻿using System;
+using UnityEngine;
```
Comment (Sunrunner37) [inline NitroxModel-Subnautica/DataStructures/DataExtensions.cs:2]: Doesn't this still require NitroxModel to take a dependency on unity?  Should this reside somewhere else?

File: NitroxModel-Subnautica/DataStructures/GameLogic/ExoSuitMovementData.cs [hunk 4]
```diff
@@ -1,31 +1,30 @@
...
         {
             LeftAimTarget = leftAimTarget;
             RightAimTarget = rightAimTarget;
         }
+
+        [ProtoMember(1)]
+        public Vector3 LeftAimTarget { get; }
+
+        [ProtoMember(2)]
+        public Vector3 RightAimTarget { get; }
```
Comment (Sunrunner37) [inline NitroxModel-Subnautica/DataStructures/GameLogic/ExoSuitMovementData.cs:28]: members moved to the bottom?

File: NitroxModel/DataStructures/GameLogic/AbsoluteEntityCell.cs [hunk 5]
```diff
@@ -50,7 +29,28 @@ public AbsoluteEntityCell(Vector3 worldSpace, int level)
...
+        public Int3 Position => BatchPosition + CellId * GetCellSize();
+
+        public Int3 Center
+        {
+            get
+            {
+                Int3 cellSize = GetCellSize();
+                return BatchPosition + CellId * cellSize + (cellSize >> 1);
+            }
         }
```
Comment (Sunrunner37) [inline NitroxModel/DataStructures/GameLogic/AbsoluteEntityCell.cs:54]: Same here.  Can we make the auto formatter less aggressive?  it is nuking a lot of git blame for no purpose.  or modify it to keep properties and members above constructor.

File: NitroxServer/Serialization/World/WorldData.cs [hunk 6]
```diff
@@ -1,12 +1,12 @@
 ﻿using System;
 using System.Collections.Generic;
-using NitroxModel.DataStructures;
 using NitroxServer.GameLogic;
 using NitroxServer.GameLogic.Bases;
 using NitroxServer.GameLogic.Entities;
 using NitroxServer.GameLogic.Items;
 using NitroxServer.GameLogic.Vehicles;
 using ProtoBufNet;
+using DTO = NitroxModel.DataStructures;
```
Comment (Sunrunner37) [inline NitroxServer/Serialization/World/WorldData.cs:9]: Can NitroxServer files directly rely on datastructures as before?  I don't see a reason here to explicitly call out DTO.  Also, we may be able to completely decouple server from unity (outside of resource parsing).

File: NitroxServer/Serialization/ServerProtobufSerializer.cs [hunk 7]
```diff
@@ -41,14 +43,15 @@ public void Deserialize(Stream stream, object o, Type t)
...
         {
-            Model.Add(typeof(UnityEngine.Light), true);
-            Model.Add(typeof(UnityEngine.BoxCollider), true);
-            Model.Add(typeof(UnityEngine.SphereCollider), true);
-            Model.Add(typeof(UnityEngine.MeshCollider), true);
-            Model.Add(typeof(UnityEngine.Vector3), false).SetSurrogate(typeof(NitroxVector3));
-            Model.Add(typeof(UnityEngine.Quaternion), false).SetSurrogate(typeof(NitroxQuaternion));
-            Model.Add(typeof(UnityEngine.Transform), false).SetSurrogate(typeof(NitroxTransform));
-            Model.Add(typeof(UnityEngine.GameObject), false).SetSurrogate(typeof(UnityStubs.GameObject));            
+            Assembly unity = AppDomain.CurrentDomain.GetAssemblies().First(assembly => assembly.GetName().Name == "UnityEngine");
```
Comment (Sunrunner37) [inline NitroxServer/Serialization/ServerProtobufSerializer.cs:46]: What is the purpose of doing this at runtime vs at compile time?  Also, aren't there now a lot unity assemblies?  Are all of these located in the same one?

File: NitroxServer/Serialization/BatchCellsParser.cs [hunk 8]
```diff
@@ -33,12 +34,12 @@ public BatchCellsParser(EntitySpawnPointFactory entitySpawnPointFactory, ServerP
             this.entitySpawnPointFactory = entitySpawnPointFactory;
             this.serializer = serializer;
 
-            surrogateTypes.Add("UnityEngine.Transform", typeof(NitroxTransform));
-            surrogateTypes.Add("UnityEngine.Vector3", typeof(NitroxVector3));
-            surrogateTypes.Add("UnityEngine.Quaternion", typeof(NitroxQuaternion));
+            // surrogateTypes.Add("UnityEngine.Transform", typeof(NitroxTransform));
+            // surrogateTypes.Add("UnityEngine.Vector3", typeof(Vector3));
+            // surrogateTypes.Add("UnityEngine.Quaternion", typeof(Quaternion));
```
Comment (Sunrunner37) [inline NitroxServer/Serialization/BatchCellsParser.cs:39]: Commented out?

File: NitroxServer/GameLogic/Unlockables/PDAStateData.cs [hunk 9]
```diff
@@ -10,10 +10,10 @@ namespace NitroxServer.GameLogic.Unlockables
     public class PDAStateData
     {
         [ProtoMember(1)]
-        public ThreadSafeCollection<TechTypeModel> UnlockedTechTypes { get; } = new ThreadSafeCollection<TechType>();
+        public ThreadSafeCollection<TechTypeModel> UnlockedTechTypes { get; } = new ThreadSafeCollection<NitroxModel.DataStructures.TechType>();
```
Comment (Sunrunner37) [inline NitroxServer/GameLogic/Unlockables/PDAStateData.cs:13]: Still referencing TechTypeModel here.

File: NitroxServer-Subnautica/GameLogic/Entities/Spawning/SlotsHelper.cs [hunk 10]
```diff
@@ -1,30 +1,31 @@
-﻿using System.Collections.Generic;
+﻿using System;
+using System.Collections.Generic;
 
 namespace NitroxServer_Subnautica.GameLogic.Entities.Spawning
```
Comment (Sunrunner37) [inline NitroxServer-Subnautica/GameLogic/Entities/Spawning/SlotsHelper.cs:4]: I think files like this we can leave untouched to preserve.

File: NitroxServer/Serialization/BatchCellsParser.cs (hunk 8 above)
Comment (Measurity) [reply NitroxServer/Serialization/BatchCellsParser.cs:39]: Working on trying to improve this.

File: NitroxServer/Serialization/ServerProtobufSerializer.cs (hunk 7 above)
Comment (Measurity) [reply NitroxServer/Serialization/ServerProtobufSerializer.cs:46]: You're correct. Didn't know how this worked previously.

File: NitroxModel-Subnautica/DataStructures/DataExtensions.cs (hunk 3 above)
Comment (Measurity) [reply NitroxModel-Subnautica/DataStructures/DataExtensions.cs:2]: This is NitroxModel-Subnautica which needs to take dependecy on Subnautica to provide a bridge between NitroxModel and programs that depend on Subnautica.

**Comment by Measurity** [comment]
I'll split some work from this PR into other PRs. Just gonna keep working on this one until everything is where I want it to be.

File: NitroxServer/Serialization/World/WorldData.cs (hunk 6 above)
Comment (Measurity) [reply NitroxServer/Serialization/World/WorldData.cs:9]: Fair enough to me. Decoupling the server is indeed the end goal. Not sure I reach it with this PR only though.

**Pull request closed without merging**